src/MongoDB.Driver/Linq/Translators/PredicateTranslator.cs

```diff
@@ -176,8 +174,14 @@ private bool CanAnyBeRenderedWithoutElemMatch(Expression node)
                 case ExpressionType.GreaterThanOrEqual:
                 case ExpressionType.LessThan:
                 case ExpressionType.LessThanOrEqual:
+                    return !wasPreviousNotEqualType;
                 case ExpressionType.NotEqual:
-                    return true;
+                    // the SERVER processes a $ne operator in a different way with
+                    // other comparison operators (see CSHARP-2012).
+                    // So, a NotEqual operator should be handled only by a "$elemMatch".
+                    // Note: The "$elemMatch" is not needed if the current node is under
+                    // a top-level NotEqual operator.
+                    return wasPreviousNotEqualType;
```

Comment:
I don't understand why we just don't return false for NotEqual.

Reply:
Let's make sure all parties understand.  @DmitryLukyanov, can you explain why?

Reply:
To show the difference please check the below Linq expression and generated queries:
_In case if we return just `false`:_
`x => x.G.Any(g => !(g.S != null))
 { "G" : { "$elemMatch" : { "S" : null } } }");`
_In case if we return `wasPreviousNotEqualType`:_
`x => x.G.Any(g => !(g.S != null))
{ "G.S" :  null }`
I can add this unit test.

Reply:
As I understand both of them are correct, but it looks like the second variant is better.

Reply:
Why is the second variant better? I think the first variant is better.

While the second variant might work correctly, it looks more like it's trying to match { ..., G : { S : null }, ... }.

If we see $elemMatch we know we are querying an array. If we see "G.S" it looks like we are querying a sub document.

Reply:
> Why is the second variant better? I think the first variant is better.
- for me, this way looks more simple without unnecessary complication, but I can be wrong.

>If we see "G.S" it looks like we are querying a subdocument. 
- as I understand this way can be used for querying both of the document elements - nested array and just subdocument. I didn't find in the documentation which way is preferable. So, I think they are equivalent.

But I don't mind to use `elemMatch` for all cases here.